File: 2021/haskell/src/Day2.hs
module Day2 (day2) where

import qualified Data.Map.Strict as Map
import Util

data Dir = Forward | Depth
  deriving (Eq, Ord)

-- >>> one <$> input
-- 1524750
one :: [(Dir, Int)] -> Int
one = foldl1' (*) . Map.elems . Map.fromListWith (+)

-- >>> two <$> input
-- 1592426537
two :: [(Dir, Int)] -> Int
two = (\(_, d, h) -> d * h)
    . foldl' (\(aim, depth, horiz) (d, i) -> case d of
                 Depth   -> (aim + i, depth, horiz)
                 Forward -> (aim, depth + aim * i, horiz + i))
             (0, 0, 0)

input :: IO [(Dir, Int)]
input = map (toDir . second read . word1) . lines
    <$> readFile "puzzle-input/day2.txt"
 where
  toDir :: (String, Int) -> (Dir, Int)
  toDir = \case
    ("forward", n) -> (Forward,   n)
    ("down"   , n) -> (Depth  ,   n)
    ("up"     , n) -> (Depth  , - n)
    _              -> error ":<"

day2 :: IO (Int, Int)
day2 = (one &&& two) <$> input

File: aoc2021/clojure-solutions/src/clojure_solutions/day2.clj
(ns clojure-solutions.day2
  (:require [clojure.string :as str]))

(defn- parse []
  (->> (slurp "./input/day2.txt")
       str/split-lines
       (map (fn [s]
              (let [[kw n] (str/split s #" ")
                    n (read-string n) ; something like ViewPatterns would be neat
                    kw (keyword kw)]
                (if (= :up kw) [:down (- n)] [kw n]))))))

;; 1524750
(defn one []
  (->> (parse)
       (reduce (fn [acc [k v]]
                 (update acc k #(+ v %)))
               {:forward 0 :down 0})
       ((fn [{f :forward d :down}]
          (* f d)))))

;; 1592426537
(defn two []
  (->> (parse)
       (reduce (fn [[aim depth horiz] [d i]]
                 (case d
                   :down    [(+ aim i) depth horiz]
                   :forward [aim (+ depth (* aim i)) (+ horiz i)]))
               [0 0 0])
       rest                             ; only depth and horiz
       (reduce *)))
